### ISSUE TITLE
怎么保存eval的结果

### ISSUE THREAD

**Issue by raymondchu-z** [PaddlePaddle/PaddleDetection#3524]
请问怎么保存eval的结果？
`--json_eval`这个接口是，是否通过已存在的bbox.json或者mask.json进行评估。而不是保存本次评估的结果。
看代码里，`COCOMetric`类下的`accumulate()`有保存eval的结果的代码，VOC格式就没有。
那VOC格式的评估结果怎么保存？

**Comment by jerrywgz** [comment]
目前只有COCO格式下会将结果保存成bbox.json或mask.json
VOC数据集通过DetectionMAP这个类统计评估结果，如果想保存voc的评估结果，需要在https://github.com/PaddlePaddle/PaddleDetection/blob/release/2.1/ppdet/engine/trainer.py#L368 这里保存输出结果和对应的gt信息

**Comment by raymondchu-z** [comment]
可以了。代码分享出来吧，虽然写的不咋地。
~~~
    def _eval_with_loader(self, loader):
        sample_num = 0
        tic = time.time()
        self._compose_callback.on_epoch_begin(self.status)
        self.status['mode'] = 'eval'
        self.model.eval()
        #按图片（data）循环，一个loader是batchsize，将数据集按batchsize分成M/batchsize个step
        # 遍历batch
        results = {'bbox': [], 'gt_bbox':[], 'gt_labels':[], 'truncated':[], 'occluded':[], 'im_file':[]}
        img_nums = len(loader.dataset.roidbs)
        results['im_file'] = [loader.dataset.roidbs[x]["im_file"] for x in range(img_nums)]
        for step_id, data in enumerate(loader):
            self.status['step_id'] = step_id
            self._compose_callback.on_step_begin(self.status)
            # forward
            outs = self.model(data)
            bboxes = outs['bbox'].numpy()
            bbox_lengths = outs['bbox_num'].numpy()
            bbox_idx = 0
            gt_boxes = data['gt_bbox']
            gt_labels = data['gt_class']
            if 'truncated' in data:
                truncated = data['truncated']
                occluded = data['occluded']
            scale_factor = data['scale_factor'].numpy(
            ) if 'scale_factor' in data else np.ones(
                (gt_boxes.shape[0], 2)).astype('float32')
            #遍历batch内的每个图片            
            for i in range(len(gt_boxes)):
                h, w = scale_factor[i]
                results['gt_bbox'].append((gt_boxes[i].numpy() / np.array([w, h, w, h])).tolist())
                results['gt_labels'] .append([int(x) for x in gt_labels[i]])
                if 'truncated' in data:
                    results['truncated'].append([int(x) for x in truncated[i]])
                    results['occluded'].append([int(x) for x in occluded[i]])
                bbox_num = bbox_lengths[i]
                bbox = bboxes[bbox_idx:bbox_idx + bbox_num]
                results['bbox'].append(bbox.tolist())
                bbox_idx += bbox_num
           
            # update metrics
            for metric in self._metrics:
                metric.update(data, outs)#, save_wrong_infer = True

            sample_num += data['im_id'].numpy().shape[0]
            self._compose_callback.on_step_end(self.status)
        output = "bbox.json"
        output = os.path.join("./json_result/", output)
        with open(output, 'w') as f:
            json.dump(results, f)
            logger.info('The bbox result is saved to bbox.json.')
~~~
